flask_resources/context.py

```diff
@@ -0,0 +1,66 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2020 CERN.
+#
+# Flask-Resources is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""Module tests."""
+
+from contextlib import contextmanager
+
+from flask import g
+from werkzeug.local import LocalProxy
+
+#
+# Proxy to the current resource context
+#
+
+
+def _get_context():
+    """Get the resource request context from the g object."""
+    if hasattr(g, "resource_requestctx"):
+        return g.resource_requestctx
+    raise RuntimeError("Working outside of resource request context.")
+
+
+resource_requestctx = LocalProxy(_get_context)
+"""Proxy to the resource's request context"""
+
+
+#
+# Resource context
+#
+
+
+class ResourceRequestCtx(object):
+    """Context manager for the resource context.
+
+    The resource request context encodes information about the currently
+    executing request for a given resource, such as:
+
+    - The mimetype selected by the content negotiation.
+    - The content type of the request payload
+    """
+
+    def __init__(
+        self,
+        accept_mimetype=None,
+        payload_mimetype=None,
+        request_args=None,
+        data=None,
+    ):
+        """Initialize the resource context."""
+        self.accept_mimetype = accept_mimetype
+        self.payload_mimetype = payload_mimetype  # Content-Type
+        self.request_args = request_args
+        self.data = data
+        self.permission_acton = None  # FIXME: Unused at the moment
+
+    def __enter__(self):
+        """Push the resource context manager on the current request."""
+        g.resource_requestctx = self
+
+    def __exit__(self, type, value, traceback):
+        """Pop the resource context manager from the current request."""
+        del g.resource_requestctx
```

Comment:
Let's chat about this in person, it might be easier. The two first points you ask are totally valid but there is no "strong" answer right now, since this is still work in progress and those are some of the questions we are trying to address. 

As for the third point, from my perspective, the "inaccessible to lower layer problem" is two folded:

a) Something is inaccesible
b) Should that something actually be needed? Or are we using wrong attributes since they belong to another scope?

Therefore:

1- First step. As you mentioned the *ResourceRequestCtx* gets into the *resource*, which is the "entity". Here is where the two first questions apply.
2- Second step. The *ResourceRequestCtx* gets striped and only certain attribtues are passed to the controller. This is done because otherwise we are just "tunneling" the problem and it will make things dependent on each other later on the road. Being the problem that something (e.g. `current_user`) should not get to the controller and only a derived of it (e.g. "user is authenticated") and all this operations should be calculated by the resouce. The controller should only be in charge of the bussiness logic (e.g. "If the user is authenticated create the record in DB, index it, etc.).

But as said let's chat about this in person, it might be easier.

Reply:
this requires more than simple comments in a PR: https://github.com/inveniosoftware/flask-resources/issues/2